sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubUnboundedSink.java

```diff
@@ -0,0 +1,321 @@
+
+  /**
+   * Publish messages to pub/sub in batches.
+   */
+  private class WriterFn
+      extends DoFn<KV<Integer, Iterable<PubsubClient.OutgoingMessage>>, Void> {
+
+    /**
+     * Client on which to talk to pub/sub. Null until created by {@link #startBundle}.
+     */
+    @Nullable
+    private transient PubsubClient pubsubClient;
+
+    private final Aggregator<Long, Long> batchCounter =
+        createAggregator("batches", new Sum.SumLongFn());
+    private final Aggregator<Long, Long> elementCounter =
+        createAggregator("elements", new Sum.SumLongFn());
+    private final Aggregator<Long, Long> byteCounter =
+        createAggregator("bytes", new Sum.SumLongFn());
+
+    /**
+     * BLOCKING
+     * Send {@code messages} as a batch to pub/sub.
+     */
+    private void publishBatch(List<PubsubClient.OutgoingMessage> messages, int bytes)
+        throws IOException {
+      long nowMsSinceEpoch = System.currentTimeMillis();
+      int n = pubsubClient.publish(topic, messages);
+      Preconditions.checkState(n == messages.size());
+      batchCounter.addValue(1L);
+      elementCounter.addValue((long) messages.size());
+      byteCounter.addValue((long) bytes);
+    }
+
+    @Override
+    public void startBundle(Context c) throws Exception {
+      Preconditions.checkState(pubsubClient == null);
+      pubsubClient = PubsubGrpcClient.newClient(timestampLabel, idLabel,
+          c.getPipelineOptions().as(GcpOptions.class));
+      super.startBundle(c);
+    }
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      List<PubsubClient.OutgoingMessage> pubsubMessages = new ArrayList<>(PUBLISH_BATCH_SIZE);
+      int bytes = 0;
+      for (PubsubClient.OutgoingMessage message : c.element().getValue()) {
+        if (!pubsubMessages.isEmpty()
+            && bytes + message.elementBytes.length > PUBLISH_BATCH_BYTES) {
+          // Break large (in bytes) batches into smaller.
+          // (We've already broken by batch size using the trigger belom, though that may
+          // run slightly over the actual PUBLISH_BATCH_SIZE)
+          // BLOCKS until published.
+          publishBatch(pubsubMessages, bytes);
+          pubsubMessages.clear();
+          bytes = 0;
+        }
+        pubsubMessages.add(message);
+        bytes += message.elementBytes.length;
+      }
+      if (!pubsubMessages.isEmpty()) {
+        // BLOCKS until published.
+        publishBatch(pubsubMessages, bytes);
+      }
+    }
+
+    @Override
+    public void finishBundle(Context c) throws Exception {
+      pubsubClient.close();
+      pubsubClient = null;
+      super.finishBundle(c);
+    }
+  }
+
+
+  // ================================================================================
+  // PubsubUnboundedSink
+  // ================================================================================
+
+  /**
+   * Number of cores available for publishing.
```

Comment:
This is now 'numShards' and needs to be chosen to balance num records batch with pubsub latency. Note that a random long indeed shards but results in most bundles having a single element which kills apiary/grpc quota.

I currently (ie in a pending branch) hard code this in PubsubIO. It is comparable to the heuristics we have baked in for calculating the initial splits for an UnboundedReader.